### PR TITLE
Time - Haben

### PR DESCRIPTION
**Pull request by HabenFoto**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Model is where our data and state for application live. It has no knowledge of UI for future usage.
Describe in your own words what the Controller is doing in Rails | The Controller receives events as requests from the user,  gathers data from the model and pass data to the view to be rendered in HTML.
Describe in your own words what the View is doing in Rails | The View turns the data received from the Controller into HTML and displays it.
Describe an edge-case controller test you wrote | N/A
What is the purpose of using strong params? (i.e. the params method in the controller) | We use strong params within our controllers to provide a safe way to require and permit data that comes from our forms.
How are Rails migrations related to Rails models? | Migrations are a convenient way for us to alter our database in a structured and organized way. Through Migration we are able to add or remove tables, columns, and rows in Model. 
Describe one area of Rails that are still unclear on | I think the fact that there are a lot of moving parts interacting with each other, is a little too much to grasp in this short amount of time. But I’m sure if I get to work on it specifically, I can have a better understanding.

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      head :not_found
+      return
+    end
+
+    task.destroy
+    redirect_to tasks_path
+   
+  end
+
+  def completed_at
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:71]: Consider naming this controller action something like `mark_complete` to avoid confusion between the `completed_at` attribute and the `completed_at` action.

Comment (beccaelenzil) [reply app/controllers/tasks_controller.rb:71]: In addition, consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: app/views/tasks/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<div>
+<h3 class="index_title">Tasks to be Completed</h3>
+<ul>
+  <% @tasks.each do |task|  %>
+    <li class="card">
+      <div class="title"><%= task.name %> <%= task.description %> <%= task.completed_at%></div>
+      <div class="buttons">
+        <%= button_to "Edit Task", edit_task_path(task[:id]), method: :get %>
+        <%= button_to "Delete Task", task_path(task[:id]), method: :delete, data: {confirm: "Are you sure you want to Delete?"} %>
+        <%= button_to "#{task.completed_at ? "Completed Task" : "Task Not Completed"}", completed_task_path(task[:id]), method: :patch %>
```
Comment (beccaelenzil) [inline app/views/tasks/index.html.erb:10]: Nice use of a ternary for conditionally showing different text based on the state of the task! Consider naming these buttons something like "Mark Complete" and "Unmark Complete" to add clarity to the user exeperience.

File: app/views/tasks/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<div>
+<h3 class="index_title">Tasks to be Completed</h3>
```
Comment (beccaelenzil) [inline app/views/tasks/index.html.erb:2]: Minor note: this title is a little misleading because you include all tasks (completed and not) in the list.

File: app/assets/stylesheets/tasks.scss [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// Place all the styles related to the Tasks controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
+@import "bootstrap";
+
+//  @import "**/*";
```
Comment (beccaelenzil) [inline app/assets/stylesheets/tasks.scss:6]: Nice work adding styling! Minor note: before you have any tasks, some of your links overlap with eachother.

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project! You're off to a great start with rails and it's clear the learning goals were met. Nice work incorporating strong_params and resources. Keep up the hard work!

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️ Good use of resources!
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
